src/bindings/wasm/tvgWasmLottieAnimation.cpp

```diff
@@ -33,46 +33,66 @@ using namespace tvg;
 
 static const char* NoError = "None";
 
-class __attribute__((visibility("default"))) TvgLottieAnimation
+struct TvgEngineMethod
 {
-public:
-    ~TvgLottieAnimation()
+    virtual ~TvgEngineMethod() {}
+    virtual unique_ptr<Canvas> init(string&) = 0;
+    virtual void resize(Canvas* canvas, int w, int h) = 0;
+    virtual val output(int w, int h)
+    {
+        return val(typed_memory_view<uint8_t>(0, nullptr));
+    }
+};
+
+struct TvgSwEngine : TvgEngineMethod
+{
+    uint8_t* buffer = nullptr;
+
+    ~TvgSwEngine()
+    {
+        free(buffer);
+        Initializer::term(tvg::CanvasEngine::Sw);
+    }
+
+    unique_ptr<Canvas> init(string&) override
+    {
+        Initializer::init(0, tvg::CanvasEngine::Sw);
+        return SwCanvas::gen();
+    }
+
+    void resize(Canvas* canvas, int w, int h) override
+    {
+        free(buffer);
+        buffer = (uint8_t*)malloc(w * h * sizeof(uint32_t));
+        static_cast<SwCanvas*>(canvas)->target((uint32_t *)buffer, w, w, h, SwCanvas::ABGR8888S);
+    }
+
+    val output(int w, int h) override
+    {
+        return val(typed_memory_view(w * h * 4, buffer));
+    }
+};
+
+
+struct TvgWgEngine : TvgEngineMethod
+{
+    #ifdef THORVG_WG_RASTER_SUPPORT
+        WGPUInstance instance{};
+        WGPUSurface surface{};
+    #endif
+
+    ~TvgWgEngine()
     {
-        if (engine == tvg::CanvasEngine::Sw) {
-        #ifdef THORVG_SW_RASTER_SUPPORT
-            free(buffer);
-        #endif
-        } else if (engine == tvg::CanvasEngine::Wg) {
         #ifdef THORVG_WG_RASTER_SUPPORT
             wgpuSurfaceRelease(surface);
             wgpuInstanceRelease(instance);
         #endif
-        }
-
-        Initializer::term();
+        Initializer::term(tvg::CanvasEngine::Wg);q
```

Comment:
q typo?

Reply:
Oops, it comes when i'm committing. Have just removed!